### DISCUSSION TITLE
Map 8-bit ASCII values

### DISCUSSION THREAD

**Discussion by huggy-d1** [houmain/keymapper#62] [Q&A] [answered]
How to map F12 to send ASCII code 0159 without hold alt, send 0159, unhold alt?
 
I have many combinations of F1-F12 (normal, shift, ctrl, alt, shift-ctrl-alt) for an old terminal emulation pkg that need remapped. 

Confirmed that using the Alt-0159 combo performs correct action on decades old computer chassis (circa 1980) for F12. 

Would prefer to have something like:
F12 >> #(0159)
Shift_F12 >> #(0179)
Ctrl-F12 >> #(0199)
Alt-F12 >> #(0209)
Shift-Ctrl-Alt >> #(0219)

What's the best way with the current software?

Thanks for an intriguing pkg.

**Comment by houmain** [comment, marked as answer]
Interesting use case! I will implement this in the next few days.

**Reply by huggy-d1** [reply]
Will your concept work for the low terminal control codes, like BEL, DC1, SOH and so on?  I can't find a text version for those codes. The input side is like ctrl-q and the output is the low ascii code values (< 32). 

Thx for so quickly responding and thinking about it. 

In the interim, the output for       F12 >> (Alt) 0 1 5 9 ;F12|BKGND
Is that correct?  Of course the preference is to simply send the code directly without making an on-screen keyboard light up with the fake key presses.

**Reply by houmain** [reply]
I am currently working on it. 
Good question, I thought adding scancode parsing to the configuration would be sufficient for your needs.
But now I see, that the keycodes < 32 are already quite common keys like Escape (see [Key.h](https://github.com/houmain/keymapper/blob/main/src/runtime/Key.h)).
On which system are you?

**Reply by houmain** [reply]
I think, this is something the terminal emulator needs to implement, don't you?

I just looked up https://en.wikipedia.org/wiki/Control_character. There it says e.g:
0x0C ([form feed](https://en.wikipedia.org/wiki/Page_break), FF, [\f](https://en.wikipedia.org/wiki/%5Cf), [^L](https://en.wikipedia.org/wiki/%5EL)), to cause a printer to eject paper to the top of the next page, or a video terminal to clear the screen.

In the keymapper config you could add this alias:
`FF = Control{ L }`
and then use it like:
`F12 >> FF`
This clears my terminal when I press F12.

**Reply by huggy-d1** [reply]
Windows 10 Enterprise 

The terminal software is emulating at VT340 just fine when we use the Alt-0159-ReleaseAlt combo. Without the leading 0, windows alt code interpreter borks the code. 

We just need to send binary value for 159 when F12 is pressed. It's called BKGND key in the CrossTalk software key-map overlay. It's a hot mess of codes. 

I've cobbled together an excel sheet to help automatically create the required strings for the .conf file once it's finalized. 

Also, I noticed # is for comments so my original suggestion to have syntax:
F12 >> #(159) ;F1:BKGND
Will not work. 

Maybe:
[Putty]
F11 >> |158| ;F11
F12 >> |159| ;F12:BKGND

**Reply by huggy-d1** [reply]
As a partial solution, just having syntax to designate raw decimal values in the .conf file is all that's needed. 

The low control codes are not used by the people using the system - AFAIK. Typically it's just variations of F1-F12. 

Those input expressions are easy to do. The output expression currently available is to press and hold Alt, the type via the NUMPAD, 0 1 5 9, then release Alt. 

Is that syntax as I previously guessed?

F12 >> (Alt) 0 1 5 9 ;F12|BKGND

If I can figure out the correct syntax for Alt code hold, enter the codes using numpad digits, release Alt, I can actually test it on Putty which has no keyboard mapping capabilities. 

It's interesting that a terminal emulator last updated in 1994 has more features than many admins go-to tool for terminal session work into remote device (Putty). 

Your tool looks awesome for lots of reasons. I've shared it with my groups newest dev to see how it might improve their day-to-day.

**Reply by huggy-d1** [reply]
> ...https://en.wikipedia.org/wiki/Control_character. ...
> 
> 0x0C ([form feed]), FF, [\f], [^L]..
> 
> In the keymapper config you could add this alias:
> 
> `FF = Control{ L }`
> 
> and then use it like:
> 
> `F12 >> FF`
> 
> This clears my terminal when I press F12.

I really like this option for the odd control codes with values <32. 

Thx

**Reply by huggy-d1** [reply]
Ok I think I found the right syntax for F-keys:

n0 = Numpad0
n1 = Numpad1
n2 = Numpad2
n3 = Numpad3
n4 = Numpad4
n5 = Numpad5
n6 = Numpad6
n7 = Numpad7
n8 = Numpad8
n9 = Numpad9

F12 >> Alt{ n0 n1 n5 n9 }

**Reply by houmain** [reply]
This is testable now in version 2.6.0.
The syntax is simply `159` or in hex notation `0x9F`.
